src/main/java/org/checkerframework/checker/startswith/StartsWithAnnotatedTypeFactory.java

```diff
@@ -0,0 +1,195 @@
+import java.lang.annotation.Annotation;
+import java.util.*;
+
+// The StartsWith checker ensures that URL strings start with the given strings.
+
+public class StartsWithAnnotatedTypeFactory extends BaseAnnotatedTypeFactory {
+
+    // The @StartsWithBottom annotation.
+    private final AnnotationMirror BOTTOM;
+
+    // The @StartsWithUnknown annotation
+    private final AnnotationMirror UNKNOWN;
+
+    private final ArrayList<String> ACCEPTED_STRINGS;
+
+    public StartsWithAnnotatedTypeFactory(BaseTypeChecker checker){
+        super(checker);
+        this.BOTTOM = AnnotationBuilder.fromClass(this.elements, StartsWithBottom.class);
+        this.UNKNOWN = AnnotationBuilder.fromClass(this.elements, StartsWithUnknown.class);
+        this.ACCEPTED_STRINGS = new ArrayList<String> (Arrays.asList("https", "file", "path"));
+        this.postInit();
+    }
+
+    @Override
+    protected void addComputedTypeAnnotations(Tree tree, AnnotatedTypeMirror type, boolean iUserFlow) {
+        this.setStringToStartsWith(tree, type);
+        this.checkConcatenatedString(tree, type);
+        super.addComputedTypeAnnotations(tree, type, iUserFlow);
+    }
+
+    @Override
+    public QualifierHierarchy createQualifierHierarchy(MultiGraphFactory factory) {
+        return new StartsWithQualifierHierarchy(factory);
+    }
+
+    @Override
+    protected Set<Class<? extends Annotation>> createSupportedTypeQualifiers() {
+            return new LinkedHashSet<>(
+                Arrays.asList(
+                        StartsWith.class,
+                        StartsWithBottom.class,
+                        StartsWithUnknown.class,
+                        PolyStartsWith.class));
+    }
+
+    /**
+     * Creates a @StartsWith annotation whose values are the given strings in a list.
+     * @param exprs
+     * @return
+     */
+    public AnnotationMirror createStartsWith(Collection<String> exprs){
+        AnnotationBuilder builder = new AnnotationBuilder(processingEnv, StartsWith.class);
+        String[] exprArray = exprs.toArray(new String[0]);
+        builder.setValue("value", exprArray);
+        return builder.build();
+    }
+
+    /**
+     * If tree is a literal tree of kind string literal and if the string starts with "https",
+     * "file" or "path" then add the StartsWith annotation so that dataflow can refine it
+     * @param tree: Given AST
+     * @param type: Type of the tree
+     */
+    private void setStringToStartsWith(Tree tree, AnnotatedTypeMirror type){
+        if(tree.getKind() == Tree.Kind.STRING_LITERAL){
+            LiteralTree literalTree = (LiteralTree) tree;
+            for(String s: this.ACCEPTED_STRINGS) {
+                if (literalTree.getValue().toString().startsWith(s)) {
+                    QualifierDefaults defaults = new QualifierDefaults(this.elements, this);
+                    defaults.addCheckedCodeDefault(createStartsWith(this.ACCEPTED_STRINGS), TypeUseLocation.ALL);
```

Comment:
When generalizing this, I expect that you'll actually just want to add a type that says the String startsWith whatever its value is. So e.g. the type of `"https"` is `@StartsWith("https")`, and the type of `"www.google.com"` is `@StartsWith("www.google.com")`.

This will also require you to change the subtyping relationship - you need `@StartsWith("https://github.com")`, for example, to be assignable to `@StartsWith("https")`.